event/event.go

```diff
@@ -105,17 +137,47 @@ func NewBus(ctx context.Context) Bus {
 }
 
 // Publish sends an event to the bus
-func (b *bus) Publish(ctx context.Context, topic Type, data interface{}) error {
+func (b *bus) Publish(ctx context.Context, topic Topic, payload interface{}) error {
+	return b.publish(ctx, topic, "", payload)
+}
+
+// Publish sends an event with a given sessionID to the bus
+func (b *bus) PublishID(ctx context.Context, topic Topic, sessionID string, payload interface{}) error {
+	return b.publish(ctx, topic, sessionID, payload)
+}
+
+func (b *bus) publish(ctx context.Context, topic Topic, sessionID string, payload interface{}) error {
 	b.lk.RLock()
 	defer b.lk.RUnlock()
-	log.Debugw("publish", "topic", topic, "payload", data)
+	log.Debugw("publish", "topic", topic, "payload", payload)
 
 	if b.closed {
 		return ErrBusClosed
 	}
 
+	e := Event{
+		Topic:     topic,
+		Timestamp: NowFunc().UnixNano(),
+		SessionID: sessionID,
+		Payload:   payload,
+	}
+
 	for _, handler := range b.subs[topic] {
-		if err := handler(ctx, topic, data); err != nil {
+		if err := handler(ctx, e); err != nil {
+			return err
+		}
+	}
+
+	if sessionID != "" {
+		for _, handler := range b.idSubs[sessionID] {
+			if err := handler(ctx, e); err != nil {
+				return err
+			}
```

Comment:
I think this bit can be a bit risky. We run the risk of double subscription. Ie a given code piece does `SubscribeAll` and in a separate step does an ID subscription to be able to isolate a given session stream. If not careful at the moment of writing we might end up with a handler being added at two points in which case weird stuff can happen.

Now as a solution I wouldn't automagically start de-duping request handlers here and would leave it to the user to keep track of what they're doing, however adding some sort of handler ID and logging it would be very useful for instrumentation and tracing now that we're going to rely so much more on the event bus. 

Thinking of it, this might not even be the right place to start solving this but maybe attach a trace to the context that's passed in?

Reply:
I think this is an issue we can skip dealing with for the time being, but I think you're right on both counts asmir: using the context param is a great place to start the instrumentation process, and having an answer for the double subscription problem would be nice to have.

For now, let's just be careful when Subscribing 😄